core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -0,0 +1,270 @@
+      this.specsById = newSpecsById;
+      return this;
+    }
+
+    Builder filterData(Expression newDataFilter) {
+      this.dataFilter = Expressions.and(dataFilter, newDataFilter);
+      return this;
+    }
+
+    Builder filterPartitions(Expression newPartitionFilter) {
+      this.partitionFilter = Expressions.and(partitionFilter, newPartitionFilter);
+      return this;
+    }
+
+    Builder caseSensitive(boolean newCaseSensitive) {
+      this.caseSensitive = newCaseSensitive;
+      return this;
+    }
+
+    Builder planWith(ExecutorService newExecutorService) {
+      this.executorService = newExecutorService;
+      return this;
+    }
+
+    DeleteFileIndex build() {
+      // read all of the matching delete manifests in parallel and accumulate the matching files in a queue
+      Queue<Pair<Integer, ManifestEntry<DeleteFile>>> deleteEntries = new ConcurrentLinkedQueue<>();
+      Tasks.foreach(deleteManifestReaders())
+          .stopOnFailure().throwFailureWhenFinished()
+          .executeWith(executorService)
+          .run(specIdAndReader -> {
+            try (CloseableIterable<ManifestEntry<DeleteFile>> reader = specIdAndReader.second()) {
+              for (ManifestEntry<DeleteFile> entry : reader) {
+                // copy with stats for better filtering against data file stats
+                deleteEntries.add(Pair.of(specIdAndReader.first(), entry.copy()));
+              }
+            } catch (IOException e) {
+              throw new RuntimeIOException("Failed to close", e);
+            }
+          });
+
+      // build a map from (specId, partition) to delete file entries
+      ListMultimap<Pair<Integer, StructLikeWrapper>, ManifestEntry<DeleteFile>> deleteFilesByPartition =
+          Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
+      for (Pair<Integer, ManifestEntry<DeleteFile>> specIdAndEntry : deleteEntries) {
+        int specId = specIdAndEntry.first();
+        ManifestEntry<DeleteFile> entry = specIdAndEntry.second();
+        deleteFilesByPartition.put(Pair.of(specId, StructLikeWrapper.wrap(entry.file().partition())), entry);
+      }
+
+      // sort the entries in each map value by sequence number and split into sequence numbers and delete files lists
+      Map<Pair<Integer, StructLikeWrapper>, Pair<long[], DeleteFile[]>> sortedDeletesByPartition = Maps.newHashMap();
+      // also, separate out equality deletes in an unpartitioned spec that should be applied globally
+      long[] globalApplySeqs = null;
+      DeleteFile[] globalDeletes = null;
+      for (Pair<Integer, StructLikeWrapper> partition : deleteFilesByPartition.keySet()) {
+        if (specsById.get(partition.first()).isUnpartitioned()) {
+          Preconditions.checkState(globalDeletes == null, "Detected multiple partition specs with no partitions");
+
+          List<Pair<Long, DeleteFile>> eqFilesSortedBySeq = deleteFilesByPartition.get(partition).stream()
```

Comment:
We iterate through delete files twice because we don't anticipate too many global delete files?

Reply:
That's right.